drivers/serial/uart_ite_it8xxx2.c

```diff
@@ -0,0 +1,959 @@
+	return TRUE;
+}
+
+/**
+ *
+ * uart_is_xmit_empty
+ *
+ * @return
+ *
+ * @parameter
+ *
+ *
+ */
+static uint8_t uart_is_xmit_empty(uint8_t chan)
+{
+	return IS_MASK_SET(mComPortTb[chan].mUartBase->ULSR, LSR_TEMT);
+}
+/**
+ *
+ * uart_init
+ *
+ * @return
+ *
+ * @parameter
+ *
+ *
+ */
+static void uart_init(uint8_t chan, uint8_t baudrate, uint8_t bINTMode,
+		      uint8_t bIsEnableParity,
+		      uint8_t bIsEvenParity,
+		      uint8_t bIsOneStopBit)
+{
+	volatile struct T_NS16550  *pComPort
+		= mComPortTb[chan].mUartBase;
+	const struct T_BaudrateStruct *pBR
+		= &mBaudrateTb[baudrate];
+	unsigned char ucTemp;
+
+	/* Enable auto gating */
+	/* UART Clock Gating (UART12CG) 0: Operation,
+	 * 1: Clocks to UART1/UART2 modules are gated.
+	 */
+	CGCTRL3R &= ~BIT2;
+	AUTOCG   &= ~(BIT(6 - chan));
+
+	/* Enable uart module */
+	RSTDMMC  |= BIT(3 - chan);      /* Set EC side control */
+	RSTC4    = BIT(1 + chan);       /* W-One to reset controller */
+	GCR1    |= mComPortTb[chan].ucGCR1;
+	GCR6    |= mComPortTb[chan].ucGCR6;
+	GCR8    |= mComPortTb[chan].ucGCR8;
+
+	if (chan == t_uart_1) {
+		GCR21 &= ~(GPH1VS | GPH2VS);
+		SET_MASK(UART2_ECSPMR, BIT1);
+		#if DEF_UART2_ENABLE_CTS_RTS    // {
+		if (mComPortTb[chan].cts) {
+			*mComPortTb[chan].cts = 0x00;
+		}
+		if (mComPortTb[chan].rts) {
+			*mComPortTb[chan].rts = 0x00;
+			SET_MASK(pComPort->UMCR, MCR_AFE);
+			ucTemp = UART2_MSR;             /* Read Clear */
+		}
+	#endif  /* DEF_UART2_ENABLE_CTS_RTS */
+	} else if (chan == t_uart_0) {
+		#if DEF_UART1_ENABLE_CTS_RTS
+		if (mComPortTb[chan].cts) {
+			*mComPortTb[chan].cts = 0x00;
+		}
+
+		if (mComPortTb[chan].rts) {
+			*mComPortTb[chan].rts = 0x00;
+			SET_MASK(pComPort->UMCR, MCR_AFE);
+			ucTemp = pComPort->UMSR;                /* Read Clear */
+		}
+		#endif /* DEF_UART1_ENABLE_CTS_RTS */
+	}
+
+	if (mComPortTb[chan].rx) {
+		*mComPortTb[chan].rx = 0x00;
+	}
+
+	if (mComPortTb[chan].tx) {
+		*mComPortTb[chan].tx = 0x00;
+	}
+
+	#if DEF_ENABLE_CTS_RTS
+	if (mComPortTb[chan].cts) {
+		*mComPortTb[chan].cts = 0x00;
+	}
+
+	if (mComPortTb[chan].rts) {
+		*mComPortTb[chan].rts = 0x00;
+		SET_MASK(pComPort->UMCR, MCR_AFE | MCR_RTS);
+		pComPort->URTR  = 56;
+	}
+	#endif /* DEF_ENABLE_CTS_RTS */
+
+	/* set baudrate */
+	#if CONFIG_ITE_UART_NS16550_DMA_MODE
+	pComPort->UBRCDIR = pBR->baud_int;
+	pComPort->UBRCDDR = pBR->baud_den;
+	pComPort->UBRCDNR = pBR->baud_num;
+	#endif /* CONFIG_ITE_UART_NS16550_DMA_MODE */
+
+	SET_MASK(pComPort->ULCR, LCR_DLAB);
+	pComPort->UDLL = pBR->baud_udll;
+	pComPort->UDLM = pBR->baud_udlm;
+	pComPort->UFCR = FCR_FIFO_EN | FCR_RFRST | FCR_XFRST |
+			 FCR_64B_FIFO_EN | FCR_FIFO_8;
+	pComPort->ULCR = LCR_8N1 | (bIsEnableParity ? LCR_PEN : 0) |
+			 (bIsEvenParity ? LCR_EPS : 0) |
+			 (bIsOneStopBit ? 0 : LCR_STB);
+
+	ISR4  |= BIT(6 + chan);
+	SET_MASK(pComPort->UMCR, MCR_OUT2);
+}
+
+
+static int get_br_idx(uint32_t baud_rate)
+{
+	switch (baud_rate) {
+	case 4800:
+		return t_br_4800;
```

Comment:
If we are going to default to the largest baud rate, should this comparison use >= intead of ==

for example 4801 would produce a baudrate of 115200, I think a better choice would be 4800 or 9600

@carlescufi, Also is there any tracing we can add for these types of error cases that we "silently fix"?

Reply:
Sure, it's more resonable that choice the closer baud rate.
We will rewrite this function:
a. Using call by reference to get baud rate index 
b. The return value will become status for tracing.

Reply:
@jettr yes, you can use `LOG()` which you should be using anyway to help out in debugging. For hard failures you can also `ASSERT()`.

Reply:
This is still unresolved. The code hasn't updated its logic (and it hasn't added LOG statement).

Reply:
We'v updated new code in this block (it became use "if", and will select the >= uart rate), or is it still bad logic? 
and we will add LOG next commit after we check the previous issue.
BR,